File: WinPaletter/Store/Store_Hover.vb
Imports WinPaletter.NativeMethods
Imports WinPaletter.XenonCore

Public Class Store_Hover

#Region "Form Shadow"

    Private aeroEnabled As Boolean

    Protected Overrides ReadOnly Property CreateParams() As CreateParams
        Get
            CheckAeroEnabled()
            Dim cp As CreateParams = MyBase.CreateParams
            If Not aeroEnabled Then
                cp.ClassStyle = cp.ClassStyle Or Dwmapi.CS_DROPSHADOW
                cp.ExStyle = cp.ExStyle Or 33554432
                Return cp
            Else
                Return cp
            End If
        End Get
    End Property

    Protected Overrides Sub WndProc(ByRef m As Message)
        Select Case m.Msg
            Case Dwmapi.WM_NCPAINT
                Dim val = 2
                If aeroEnabled Then
                    Dwmapi.DwmSetWindowAttribute(Handle, If(GetRoundedCorners(), 2, 1), val, 4)
                    Dim bla As New Dwmapi.MARGINS()
                    With bla
                        .bottomHeight = 1
                        .leftWidth = 1
                        .rightWidth = 1
                        .topHeight = 1
                    End With
                    Dwmapi.DwmExtendFrameIntoClientArea(Handle, bla)
                End If
                Exit Select
        End Select

        Const WM_NCACTIVATE As UInt32 = &H86

        If m.Msg = WM_NCACTIVATE AndAlso m.WParam.ToInt32() = 0 Then
            HandleDeactivate()
        End If

        MyBase.WndProc(m)
    End Sub

    Private Sub CheckAeroEnabled()
        If Environment.OSVersion.Version.Major >= 6 Then
            Dim Com As Boolean
            Dwmapi.DwmIsCompositionEnabled(Com)
            aeroEnabled = Com
        Else
            aeroEnabled = False
        End If
    End Sub

    Sub HandleDeactivate()
        If _shown Then
            _shown = False
            Close()
        End If
    End Sub
#End Region

    Private index As Integer = 0
    Public img0 As Bitmap
    Public img1 As Bitmap
    Private _shown As Boolean = False

    Private Sub Store_Hover_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Icon = Store.Icon
        LoadLanguage
        ApplyDarkMode(Me)
        DoubleBuffer

        Dim p As Point = Store.selectedItem.PointToScreen(Point.Empty)
        Location = p

        _shown = False
    End Sub

    Private Sub Store_Hover_Shown(sender As Object, e As EventArgs) Handles Me.Shown
        _shown = True
    End Sub

    Private Sub Store_Hover_MouseClick(sender As Object, e As MouseEventArgs) Handles Me.MouseClick
        Close()
    End Sub

    Private Sub Store_Hover_KeyUp(sender As Object, e As KeyEventArgs) Handles Me.KeyUp
        If e.KeyValue = Keys.Right Or e.KeyValue = Keys.Left Or e.KeyValue = Keys.Up Or e.KeyValue = Keys.Down Then
            SwitchPreview()
        Else
            Close()
        End If
    End Sub

    Private Sub Store_Hover_MouseWheel(sender As Object, e As MouseEventArgs) Handles Me.MouseWheel
        SwitchPreview()
    End Sub

    Sub SwitchPreview()
        If index = 0 Then
            index = 1
            If img1 IsNot Nothing Then BackgroundImage = img1
        Else
            index = 0
            If img0 IsNot Nothing Then BackgroundImage = img0
        End If
    End Sub

    Private Sub Store_Hover_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed
        My.RenderingHint = If(My.CP.MetricsFonts.Fonts_SingleBitPP, Drawing.Text.TextRenderingHint.SingleBitPerPixelGridFit, Drawing.Text.TextRenderingHint.ClearTypeGridFit)
    End Sub
End Class

File: WinPaletter/Store/Store_Hover.cs
using System;
using System.Drawing;
using System.Drawing.Text;
using System.Windows.Forms;

namespace WinPaletter
{
    /// <summary>
    /// Hover form for previewing themes in the store.
    /// </summary>
    public partial class Store_Hover
    {

        /// <summary>
        /// Initializes a new instance of the <see cref="Store_Hover"/> class.
        /// </summary>
        public Store_Hover()
        {
            InitializeComponent();
        }

        private int index = 0;

        /// <summary>
        /// Image to display in the preview (modern Windows elements).
        /// </summary>
        public Bitmap img0;

        /// <summary>
        /// Image to display in the preview (classic Windows elements).
        /// </summary>
        public Bitmap img1;

        private void Store_Hover_Load(object sender, EventArgs e)
        {
            Point p;
            if (Forms.Store.selectedItem is not null)
            {
                p = Forms.Store.selectedItem.PointToScreen(Point.Empty) - (Size)new Point((int)((Width - Forms.Store.selectedItem.Width) / 2f), (int)((Height - Forms.Store.selectedItem.Height) / 2f));
            }
            else
            {
                p = MousePosition;
            }

            if (p.X + Width > Screen.PrimaryScreen.Bounds.Width) p = new(Screen.PrimaryScreen.Bounds.Width - Width, p.Y);
            if (p.Y + Height > Screen.PrimaryScreen.Bounds.Height) p = new(p.X, Screen.PrimaryScreen.Bounds.Height - Height);
            Location = p;
        }

        private void Store_Hover_KeyUp(object sender, KeyEventArgs e)
        {
            if (e.KeyValue == (int)Keys.Right | e.KeyValue == (int)Keys.Left | e.KeyValue == (int)Keys.Up | e.KeyValue == (int)Keys.Down)
            {
                SwitchPreview();
            }
            else
            {
                Close();
            }
        }

        private void Store_Hover_MouseWheel(object sender, MouseEventArgs e)
        {
            SwitchPreview();
        }

        /// <summary>
        /// Switches the preview image between the two available images (modern and classic Windows elements).
        /// </summary>
        private void SwitchPreview()
        {
            if (index == 0)
            {
                index = 1;
                if (img1 is not null) BackgroundImage = img1;
            }
            else
            {
                index = 0;
                if (img0 is not null) BackgroundImage = img0;
            }
        }

        private void Store_Hover_FormClosed(object sender, FormClosedEventArgs e)
        {
            Program.Style.TextRenderingHint = Program.TM.MetricsFonts.Fonts_SingleBitPP ? TextRenderingHint.SingleBitPerPixelGridFit : TextRenderingHint.ClearTypeGridFit;
        }
    }
}